File: php/features.php
<?php
declare(strict_types=1);

// LmMultichannel SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class LmMultichannelFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new LmMultichannelBaseFeature();
            case "debug":
                return new LmMultichannelDebugFeature();
            case "idempotency":
                return new LmMultichannelIdempotencyFeature();
            case "metrics":
                return new LmMultichannelMetricsFeature();
            case "paging":
                return new LmMultichannelPagingFeature();
            case "ratelimit":
                return new LmMultichannelRatelimitFeature();
            case "retry":
                return new LmMultichannelRetryFeature();
            case "test":
                return new LmMultichannelTestFeature();
            case "timeout":
                return new LmMultichannelTimeoutFeature();
            default:
                return new LmMultichannelBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "debug":
            case "idempotency":
            case "metrics":
            case "paging":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}

File: rb/features.rb
# LmMultichannel SDK feature factory

require_relative 'feature/base_feature'
require_relative 'feature/debug_feature'
require_relative 'feature/idempotency_feature'
require_relative 'feature/metrics_feature'
require_relative 'feature/paging_feature'
require_relative 'feature/ratelimit_feature'
require_relative 'feature/retry_feature'
require_relative 'feature/test_feature'
require_relative 'feature/timeout_feature'


module LmMultichannelFeatures
  def self.make_feature(name)
    case name
    when "base"
      LmMultichannelBaseFeature.new
    when "debug"
      LmMultichannelDebugFeature.new
    when "idempotency"
      LmMultichannelIdempotencyFeature.new
    when "metrics"
      LmMultichannelMetricsFeature.new
    when "paging"
      LmMultichannelPagingFeature.new
    when "ratelimit"
      LmMultichannelRatelimitFeature.new
    when "retry"
      LmMultichannelRetryFeature.new
    when "test"
      LmMultichannelTestFeature.new
    when "timeout"
      LmMultichannelTimeoutFeature.new
    else
      LmMultichannelBaseFeature.new
    end
  end
end
